File: 4128-total-waviness-of-numbers-in-range-ii/total-waviness-of-numbers-in-range-ii.py
class Solution:
    def totalWaviness(self, num1: int, num2: int) -> int:
        def solve(n: int) -> int:
            if n <= 0:
                return 0
            s = str(n)
            memo = {}
            
            def dp(idx, is_tight, is_lz, pp_d, p_d):
                if idx == len(s):
                    return 1, 0
                state = (idx, is_lz, pp_d, p_d)
                if not is_tight and state in memo:
                    return memo[state]
                
                limit = int(s[idx]) if is_tight else 9
                t_cnt, t_wav = 0, 0
                
                for d in range(limit + 1):
                    n_tight = is_tight and d == limit
                    n_lz = is_lz and d == 0
                    n_pp = -1 if n_lz else p_d
                    n_p = -1 if n_lz else d
                    
                    add_wav = 0
                    if pp_d != -1 and p_d != -1:
                        if (pp_d < p_d and p_d > d) or (pp_d > p_d and p_d < d):
                            add_wav = 1
                            
                    c, w = dp(idx + 1, n_tight, n_lz, n_pp, n_p)
                    t_cnt += c
                    t_wav += w + add_wav * c
                
                if not is_tight:
                    memo[state] = (t_cnt, t_wav)
                return t_cnt, t_wav
                
            return dp(0, True, True, -1, -1)[1]

        return solve(num2) - solve(num1 - 1)

File: 4128-total-waviness-of-numbers-in-range-ii/total-waviness-of-numbers-in-range-ii.java
class Solution {
    public long totalWaviness(long num1, long num2) {
        return solve(num2) - solve(num1 - 1);
    }

    private long solve(long n) {
        if (n <= 0) return 0;
        String s = String.valueOf(n);
        long[][][][][] memo = new long[s.length()][2][12][12][2];
        
        for (int i = 0; i < s.length(); i++) {
            for (int j = 0; j < 2; j++) {
                for (int k = 0; k < 12; k++) {
                    for (int l = 0; l < 12; l++) {
                        memo[i][j][k][l][0] = -1;
                    }
                }
            }
        }
        
        return dp(s, 0, true, true, -1, -1, memo)[1];
    }

    private long[] dp(String s, int idx, boolean isTight, boolean isLz, int ppD, int pD, long[][][][][] memo) {
        if (idx == s.length()) {
            return new long[]{1, 0};
        }
        
        int mlz = isLz ? 1 : 0;
        int mpp = ppD + 1;
        int mp = pD + 1;

        if (!isTight && memo[idx][mlz][mpp][mp][0] != -1) {
            return memo[idx][mlz][mpp][mp];
        }

        int limit = isTight ? s.charAt(idx) - '0' : 9;
        long tCnt = 0;
        long tWav = 0;

        for (int d = 0; d <= limit; d++) {
            boolean nTight = isTight && d == limit;
            boolean nLz = isLz && d == 0;
            int nPp = nLz ? -1 : pD;
            int nP = nLz ? -1 : d;

            int addWav = 0;
            if (ppD != -1 && pD != -1) {
                if ((ppD < pD && pD > d) || (ppD > pD && pD < d)) {
                    addWav = 1;
                }
            }

            long[] res = dp(s, idx + 1, nTight, nLz, nPp, nP, memo);
            tCnt += res[0];
            tWav += res[1] + addWav * res[0];
        }

        long[] ans = new long[]{tCnt, tWav};
        if (!isTight) {
            memo[idx][mlz][mpp][mp] = ans;
        }
        return ans;
    }
}